lib/src/stream_sink_transformer/reject_errors.dart

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+/// A [StreammSink] wrapper that rejects all errors passed into the sink.
+class RejectErrorsSink<T> implements StreamSink<T> {
+  /// The inner sink being wrapped.
+  final StreamSink<T> _inner;
+
+  @override
+  Future<void> get done => _doneCompleter.future;
+  final _doneCompleter = Completer();
+
+  /// Whether the underlying sink is no longer receiving events.
+  ///
+  /// This can happen if:
+  ///
+  /// * [close] has been called,
+  /// * an error has been passed,
+  /// * or the underlying [StreamSink.done] has completed.
+  ///
+  /// If [_canceled] is true, [_inAddStream] must be false.
+  bool get _canceled => _doneCompleter.isCompleted;
+
+  /// Whether the user has called [close].
+  ///
+  /// If [_closed] is true, [_canceled] must be true and [_inAddStream] must be
+  /// false.
+  bool _closed = false;
+
+  /// The subscription to the stream passed to [addStream], if a stream is
+  /// currently being added.
+  StreamSubscription<T>? _addStreamSubscription;
+
+  /// The completer for the future returned by [addStream], if a stream is
+  /// currently being added.
+  Completer<void>? _addStreamCompleter;
+
+  /// Whether we're currently adding a stream with [addStream].
+  bool get _inAddStream => _addStreamSubscription != null;
+
+  RejectErrorsSink(this._inner) {
+    _inner.done.whenComplete(() {
+      if (_inAddStream) {
+        _addStreamCompleter!.complete(_addStreamSubscription!.cancel());
+        _addStreamCompleter = null;
+        _addStreamSubscription = null;
+      }
+    }).then((value) {
+      if (!_canceled) _doneCompleter.complete(value);
+    }).onError<Object>((error, stackTrace) {
```

Comment:
While I applaud the use of `onError`, I'd just combine it into one `then` call with an `onError` argument, instead of creating an unnecessary intermediate future.
So, all in all:
```dart
  _inner.done.then((value) {
    _cancelAddStream();
    if (!_canceled) _doneCompleter.complete(value);
  }, onError: (error, stackTrace) {
    _cancelAddStream();
    if (!_canceled) _doneCompleter.completeError(error, stackTrace);
  });
```

Reply:
I did that at first, but it doesn't work well with the type system—the `onError` parameter just takes a `Function`, so all the parameters are `dynamic` and would require explicit casts to pass into `Completer.completeError()`.

Reply:
True, one should type the parameters in that case:
```dart
}, onError: (Objeft error, StackTrace stackTrace) {
    _cancelAddStream();
    if (!_canceled) _doneCompleter.completeError(error, stackTrace);
});
```
No casting needed then.